aten/src/ATen/native/LossNLL.cpp

```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
   return grad_input;
 }
 
+Tensor cross_entropy_loss_prob_target(
+    const Tensor& self,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction) {
+  const auto n_classes = self.size(1);
+  TORCH_CHECK(
+      !weight.defined() || (weight.dim() == 1 && weight.numel() == n_classes),
+      "weight tensor should be defined either for all ",
+      n_classes,
+      " classes or no classes"
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  Tensor ret;
+  Tensor input = at::log_softmax(
+      self, 1, optTypeMetaToScalarType(self.options().dtype_opt()));
+  if (weight.defined()) {
+    // Expand weight to the correct number of dims for broadcasting with input / target
+    auto weight_bc_shape = std::vector<int64_t>(input.dim(), 1);
+    weight_bc_shape[1] = weight.size(0);
+    Tensor weight_ = weight.view(weight_bc_shape);
+
+    ret = -(input * target * weight_).sum(1);
+    if (reduction == Reduction::Mean) {
+      // Compute weighted mean
+      ret = ret.sum() / (target * weight_).sum();
+    } else if (reduction == Reduction::Sum) {
+      ret = ret.sum();
+    }
+  } else {
+    ret = -(input * target).sum(1);
+    if (reduction == Reduction::Mean) {
+      ret = ret.mean();
+    } else if (reduction == Reduction::Sum) {
+      ret = ret.sum();
+    }
+  }
+  return ret;
+}
+
 Tensor cross_entropy_loss(
     const Tensor& self,
     const Tensor& target,
     const c10::optional<Tensor>& weight,
     int64_t reduction,
     int64_t ignore_index) {
-  return at::nll_loss_nd(
-      at::log_softmax(
-          self, 1, optTypeMetaToScalarType(self.options().dtype_opt())),
-      target,
-      weight,
-      reduction,
-      ignore_index);
+  Tensor ret;
```

Comment:
nit: you can just return inside each of the branches of the conditional to save two lines of code

Reply:
does this break NRVO though?